File: sdks/python/src/fireweave/adapters/remote.py
"""Fireweave remote backend adapter (ADR-0005) — default production path.

HTTP client for fw-server ``POST /v1/flags/evaluate`` and ``POST /v1/capture``.
Auth: ``Authorization: Bearer <FW_PROJECT_API_KEY>``. No PostHog dependency.
"""

from __future__ import annotations

import json
import os
import threading
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
from urllib.parse import urlparse

from ..config import FireweaveConfig
from ..context import EvaluationContext
from ..errors import (
    AlreadyClosedError,
    AuthenticationError,
    AuthorizationError,
    BackendUnavailableError,
    ConfigurationError,
    InvalidContextError,
    MalformedResponseError,
    NetworkError,
    NotReadyError,
    RateLimitedError,
    TimeoutError_,
)
from .base import FlagResolution

__all__ = ["FireweaveRemoteAdapter"]

_EVALUATE_PATH = "/v1/flags/evaluate"
_CAPTURE_PATH = "/v1/capture"


@dataclass
class FireweaveRemoteAdapter:
    """Vendor-neutral remote adapter speaking the Fireweave wire protocol."""

    api_url: Optional[str] = None
    api_key: Optional[str] = None
    allowed_hosts: Optional[tuple[str, ...]] = None
    request_timeout_ms: int = 3000
    shutdown_timeout_ms: int = 10000
    # Test injection: callable(url, data, headers, timeout) -> (status, body_dict)
    transport: Any = None

    backend_name: str = "fireweave"
    _ready: bool = field(default=False, init=False, repr=False)
    _closed: bool = field(default=False, init=False, repr=False)
    _pending: List[Dict[str, Any]] = field(default_factory=list, init=False, repr=False)
    _lock: threading.Lock = field(default_factory=threading.Lock, init=False, repr=False)

    def initialize(self, config: Optional[FireweaveConfig] = None) -> None:
        if self._closed:
            raise AlreadyClosedError()
        api_url = (self.api_url or os.environ.get("FW_API_URL") or "").rstrip("/")
        api_key = self.api_key or os.environ.get("FW_PROJECT_API_KEY") or ""
        if config is not None:
            if not api_url and config.host:
                api_url = config.host.rstrip("/")
            if not api_key and config.project_api_key:
                api_key = config.project_api_key
            if config.feature_flags_request_timeout_ms:
                self.request_timeout_ms = config.feature_flags_request_timeout_ms
        if not api_url or not api_key:
            raise ConfigurationError("invalid configuration", init_fatal=True)
        parsed = urlparse(api_url)
        hostname = (parsed.hostname or "").lower()
        if parsed.scheme not in ("http", "https") or not hostname:
            raise ConfigurationError("invalid configuration", init_fatal=True)
        if parsed.scheme == "http" and hostname not in ("localhost", "127.0.0.1", "::1"):
            raise ConfigurationError("invalid configuration", init_fatal=True)
        allow = self.allowed_hosts
        if allow is None and config is not None and config.allowed_hosts is not None:
            allow = config.allowed_hosts
        if allow is None:
            allow = (hostname, "localhost", "127.0.0.1", "::1")
        if "*" not in allow and hostname not in {h.lower() for h in allow}:
            raise ConfigurationError("invalid configuration", init_fatal=True)
        self.api_url = api_url
        self.api_key = api_key
        self._ready = True

    def resolve(self, flag_key: str, context: EvaluationContext) -> FlagResolution:
        if self._closed:
            raise AlreadyClosedError()
        if not self._ready:
            raise NotReadyError()
        targeting = context.targeting_key or ""
        if not targeting:
            raise InvalidContextError("targeting key missing")

        attributes: Dict[str, Any] = {}
        groups = None
        group_properties = None
        for k, v in (context.attributes or {}).items():
            if k in ("groups", "fireweave.groups") and isinstance(v, dict):
                groups = v
                continue
            if k in ("groupProperties", "fireweave.groupProperties") and isinstance(
                v, dict
            ):
                group_properties = v
                continue
            if k.startswith("$") or k.startswith("fireweave."):
                continue
            attributes[k] = v

        body: Dict[str, Any] = {"targetingKey": targeting, "flagKeys": [flag_key]}
        if attributes:
            body["attributes"] = attributes
        if groups:
            body["groups"] = groups
        if group_properties:
            body["groupProperties"] = group_properties

        data = self._request(_EVALUATE_PATH, body)
        decisions = data.get("decisions") or []
        item = next((d for d in decisions if d.get("flagKey") == flag_key), None)
        if item is None or item.get("found") is False:
            return FlagResolution(
                value=None,
                matched=False,
                enabled=False,
                quota_limited=bool(data.get("quotaLimited")),
            )
        meta = item.get("flagMetadata") or {}
        return FlagResolution(
            value=item.get("value"),
            variant=item.get("variant"),
            enabled=bool(item.get("enabled", True)),
            matched=True,
            version=meta.get("fireweave.flagVersion"),
            vendor_flag_id=meta.get("fireweave.vendorFlagId"),
            reason_code=meta.get("fireweave.reasonCode"),
            payload=item.get("payload"),
            fireweave_reason=item.get("reason"),
            quota_limited=bool(data.get("quotaLimited") or meta.get("fireweave.quotaLimited")),
        )

    def send_exposures(self, events: list) -> None:
        if self._closed or not self._ready:
            return
        with self._lock:
            for ev in events:
                if isinstance(ev, dict):
                    self._pending.append(
                        {
                            "type": "exposure",
                            "targetingKey": ev.get("targetingKey") or ev.get("targeting_key") or "",
                            "flagKey": ev.get("flagKey") or ev.get("flag_key"),
                            "value": ev.get("value"),
                            "variant": ev.get("variant"),
                        }
                    )

    def deliver_signal(self, signal: Dict[str, Any]) -> None:
        if self._closed or not self._ready:
            return
        with self._lock:
            self._pending.append(
                {
                    "type": "signal",
                    "targetingKey": signal.get("targetingKey")
                    or signal.get("targeting_key")
                    or "fireweave-sdk",
                    "name": signal.get("name"),
                    "flagKey": signal.get("flagKey") or signal.get("flag_key"),
                    "variant": signal.get("variant"),
                    "properties": {
                        "kind": signal.get("kind"),
                        "status": signal.get("status"),
                    },
                }
            )

    def flush(self) -> None:
        if self._closed or not self._ready:
            return
        with self._lock:
            batch = list(self._pending)
            self._pending.clear()
        if not batch:
            return
        try:
            self._request(_CAPTURE_PATH, {"events": batch})
        except Exception:
            with self._lock:
                self._pending[:0] = batch

    def shutdown(self, timeout_ms: int) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            self.flush()
        except Exception:
            pass
        self._ready = False

    def runtime_features(self) -> Dict[str, bool]:
        return {
            "remoteEvaluation": True,
            "localEvaluation": False,
            "localOnly": False,
            "exposureEmission": True,
            "sideEffectFreeReads": True,
            "groupAnalytics": True,
        }

    def _request(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        url = f"{self.api_url}{path}"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}",
        }
        timeout = max(self.request_timeout_ms / 1000.0, 0.001)
        if self.transport is not None:
            status, parsed = self.transport(url, body, headers, timeout)
        else:
            req = urllib.request.Request(
                url,
                data=json.dumps(body).encode("utf-8"),
                headers=headers,
                method="POST",
            )
            try:
                with urllib.request.urlopen(req, timeout=timeout) as resp:
                    status = resp.status
                    raw = resp.read().decode("utf-8")
            except urllib.error.HTTPError as e:
                status = e.code
                raw = e.read().decode("utf-8") if e.fp else ""
            except TimeoutError as e:
                raise TimeoutError_() from e
            except urllib.error.URLError as e:
                reason = getattr(e, "reason", e)
                if isinstance(reason, TimeoutError) or "timed out" in str(reason).lower():
                    raise TimeoutError_() from e
                raise NetworkError() from e
            try:
                parsed = json.loads(raw) if raw else {}
            except json.JSONDecodeError as e:
                raise MalformedResponseError() from e
        if status == 401:
            raise AuthenticationError()
        if status == 403:
            raise AuthorizationError()
        if status == 429:
            raise RateLimitedError()
        if status >= 500:
            raise BackendUnavailableError()
        if status >= 400:
            raise BackendUnavailableError()
        if not isinstance(parsed, dict):
            raise MalformedResponseError()
        return parsed

File: sdks/node/packages/sdk/src/adapters/remote.ts
/**
 * Fireweave remote backend adapter (ADR-0005) — **default production path**.
 *
 * Speaks only the vendor-neutral Fireweave remote protocol to fw-server:
 *   POST /v1/flags/evaluate
 *   POST /v1/capture
 *   POST /v1/targets/register
 *
 * Auth: Authorization: Bearer <FW_PROJECT_API_KEY> (project-api-key_…).
 * Which backend fw-server forwards to is fw-server's concern: no vendor SDK,
 * key, or host ever enters the application process.
 *
 * See spec/remote-protocol.md.
 */
import { readEnv } from '../env.js';
import { FireweaveError } from '../errors.js';
import { assertHostAllowed, isLoopbackHostname } from '../hosts.js';
import { validateTargetingKey } from '../validation.js';
import type {
  AdapterResolution,
  AdapterRuntimeFeatures,
  BackendAdapter,
  RegisterTargetOptions,
  RegisterTargetResult,
  ResolveOptions,
} from '../adapter.js';
import type { CanonicalContext, Exposure, JsonValue, Signal } from '../types.js';

const DEFAULT_ADAPTER_SHUTDOWN_TIMEOUT_MS = 10000;
const DEFAULT_REQUEST_TIMEOUT_MS = 3000;
const EVALUATE_PATH = '/v1/flags/evaluate';
const CAPTURE_PATH = '/v1/capture';
const REGISTER_TARGET_PATH = '/v1/targets/register';

type FetchLike = (
  url: string,
  init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: string;
    signal?: AbortSignal;
  },
) => Promise<{
  status: number;
  text(): Promise<string>;
  json(): Promise<unknown>;
}>;

export interface FireweaveRemoteAdapterOptions {
  /**
   * fw-server base URL (e.g. https://fw.example.com or http://127.0.0.1:3901).
   * Env alias: FW_API_URL.
   */
  apiUrl?: string;
  /**
   * Fireweave project/runtime key (project-api-key_…).
   * Env alias: FW_PROJECT_API_KEY.
   */
  apiKey?: string;
  /**
   * SSRF allowlist override. Default: hostname of `apiUrl` + loopback.
   * Pass `['*']` to opt out of host pinning.
   */
  allowedHosts?: readonly string[];
  /** Per-request timeout for evaluate/capture (default 3000 ms). */
  requestTimeoutMs?: number;
  /** Deadline for flush during shutdown (default 10 000 ms). */
  shutdownTimeoutMs?: number;
  /** Injected fetch (tests). */
  fetch?: FetchLike;
}

interface DecisionItem {
  flagKey: string;
  value: JsonValue;
  reason: string;
  found: boolean;
  enabled?: boolean;
  variant?: string | null;
  payload?: JsonValue;
  flagMetadata?: Record<string, string | number | boolean>;
}

interface EvaluateResponse {
  decisions?: DecisionItem[];
  requestId?: string;
  quotaLimited?: boolean;
}

interface CaptureEvent {
  type: 'exposure' | 'signal' | 'event';
  targetingKey: string;
  name?: string;
  flagKey?: string;
  value?: JsonValue;
  variant?: string | null;
  timestamp?: string;
  rolloutId?: string;
  changeId?: string;
  stampId?: string;
  properties?: Record<string, JsonValue>;
}

function mapHttpStatus(status: number): FireweaveError {
  if (status === 401) return new FireweaveError('Authentication');
  if (status === 403) return new FireweaveError('Authorization');
  if (status === 429) return new FireweaveError('RateLimited');
  if (status >= 500) return new FireweaveError('BackendUnavailable');
  return new FireweaveError('BackendUnavailable');
}

function resolveFromEnv(options: FireweaveRemoteAdapterOptions): {
  apiUrl: string;
  apiKey: string;
} {
  const apiUrl = (options.apiUrl ?? readEnv('FW_API_URL') ?? '').replace(/\/+$/, '');
  const apiKey = options.apiKey ?? readEnv('FW_PROJECT_API_KEY') ?? '';
  return { apiUrl, apiKey };
}

function defaultAllowedHostsFor(apiUrl: string): string[] {
  try {
    const hostname = new URL(apiUrl).hostname;
    const hosts = [hostname, 'localhost', '127.0.0.1', '::1'];
    if (hostname === '[::1]') hosts.push('::1');
    return hosts;
  } catch {
    return ['localhost', '127.0.0.1', '::1'];
  }
}

function reasonToAdapter(reason: string): AdapterResolution['reason'] | undefined {
  const allowed = new Set([
    'STATIC',
    'DEFAULT',
    'TARGETING_MATCH',
    'SPLIT',
    'CACHED',
    'STALE',
    'DISABLED',
    'UNKNOWN',
    'ERROR',
  ]);
  return allowed.has(reason) ? (reason as AdapterResolution['reason']) : undefined;
}

export class FireweaveRemoteAdapter implements BackendAdapter {
  readonly name = 'fireweave' as const;
  private readonly options: FireweaveRemoteAdapterOptions;
  private apiUrl = '';
  private apiKey = '';
  private closed = false;
  private ready = false;
  private readonly pending: CaptureEvent[] = [];
  private readonly fetchImpl: FetchLike;

  constructor(options: FireweaveRemoteAdapterOptions = {}) {
    this.options = { ...options };
    this.fetchImpl = options.fetch ?? (globalThis.fetch as unknown as FetchLike);
  }

  async initialize(_signal?: AbortSignal): Promise<void> {
    if (this.closed) throw new FireweaveError('AlreadyClosed');
    const { apiUrl, apiKey } = resolveFromEnv(this.options);
    if (!apiUrl || !apiKey) {
      throw new FireweaveError('Configuration');
    }
    let url: URL;
    try {
      url = new URL(apiUrl);
    } catch {
      throw new FireweaveError('Configuration');
    }
    if (url.protocol !== 'http:' && url.protocol !== 'https:') {
      throw new FireweaveError('Configuration');
    }
    if (url.protocol === 'http:' && !isLoopbackHostname(url.hostname)) {
      throw new FireweaveError('Configuration');
    }
    const allow =
      this.options.allowedHosts !== undefined && this.options.allowedHosts.length > 0
        ? this.options.allowedHosts
        : defaultAllowedHostsFor(apiUrl);
    assertHostAllowed(apiUrl, allow);
    this.apiUrl = apiUrl.replace(/\/+$/, '');
    this.apiKey = apiKey;
    this.ready = true;
  }

  async resolve(
    flagKey: string,
    context: CanonicalContext,
    options?: ResolveOptions,
  ): Promise<AdapterResolution> {
    if (this.closed) throw new FireweaveError('AlreadyClosed');
    if (!this.ready) throw new FireweaveError('NotReady');
    const targetingKeyResult = validateTargetingKey(context.targetingKey, true);
    if (!targetingKeyResult.ok) throw targetingKeyResult.error;
    const targetingKey = targetingKeyResult.value ?? '';

    const attributes: Record<string, JsonValue> = {};
    for (const [k, v] of Object.entries(context.attributes)) {
      if (k === 'groups' || k === 'groupProperties' || k.startsWith('$') || k.startsWith('fireweave.')) {
        continue;
      }
      attributes[k] = v;
    }

    const body: Record<string, unknown> = {
      targetingKey,
      flagKeys: [flagKey],
    };
    if (Object.keys(attributes).length > 0) body['attributes'] = attributes;
    if (context.groups !== undefined) body['groups'] = context.groups;
    if (context.groupProperties !== undefined) body['groupProperties'] = context.groupProperties;

    const response = await this.requestJson<EvaluateResponse>(
      EVALUATE_PATH,
      body,
      options?.signal,
    );
    const item = response.decisions?.find((d) => d.flagKey === flagKey);
    if (item === undefined) {
      const missing: AdapterResolution = { found: false };
      if (response.quotaLimited === true) missing.quotaLimited = true;
      return missing;
    }
    return this.toResolution(item, response.quotaLimited === true);
  }

  /**
   * Register a user or device so flag rules can target its DURABLE properties.
   *
   * Call once per login / device provisioning, then send the same
   * `targetingKey` on evaluate. Per-request `attributes` still override the
   * stored properties for a single evaluation — the two identity paths compose
   * (see spec/remote-protocol.md § Two identity paths).
   *
   * Never throws for transport failures: registration sits in login paths, and
   * an analytics call must not break sign-in. Retried ONCE when the error
   * taxonomy marks the failure retryable (network / timeout / backend); a
   * rejected payload or bad key is not retried, since it would be rejected
   * identically. The result object reports what happened for callers that
   * want to log it.
   */
  async registerTarget(
    targetingKey: string,
    options: RegisterTargetOptions = {},
  ): Promise<RegisterTargetResult> {
    if (this.closed) return { ok: false, error: new FireweaveError('AlreadyClosed') };
    if (!this.ready) return { ok: false, error: new FireweaveError('NotReady') };
    const targetingKeyResult = validateTargetingKey(targetingKey, true);
    if (!targetingKeyResult.ok) return { ok: false, error: targetingKeyResult.error };

    const body: Record<string, unknown> = { targetingKey };
    if (options.kind !== undefined) body['kind'] = options.kind;
    if (options.environment !== undefined) body['environment'] = options.environment;
    if (options.properties !== undefined && Object.keys(options.properties).length > 0) {
      body['properties'] = options.properties;
    }

    let lastError: FireweaveError | undefined;
    for (let attempt = 0; attempt < 2; attempt += 1) {
      try {
        await this.requestJson(REGISTER_TARGET_PATH, body, options.signal);
        return { ok: true };
      } catch (err) {
        lastError =
          err instanceof FireweaveError ? err : new FireweaveError('BackendUnavailable');
        if (!lastError.retryable) break;
      }
    }
    return { ok: false, ...(lastError !== undefined ? { error: lastError } : {}) };
  }

  recordExposure(exposure: Exposure): void {
    if (this.closed || !this.ready) return;
    const event: CaptureEvent = {
      type: 'exposure',
      targetingKey: exposure.targetingKey,
      flagKey: exposure.flagKey,
      value: exposure.value,
    };
    if (exposure.variant !== undefined) event.variant = exposure.variant;
    if (exposure.rolloutId !== undefined) event.rolloutId = exposure.rolloutId;
    if (exposure.changeId !== undefined) event.changeId = exposure.changeId;
    if (exposure.stampId !== undefined) event.stampId = exposure.stampId;
    this.pending.push(event);
  }

  recordSignal(signal: Signal): void {
    if (this.closed || !this.ready) return;
    const event: CaptureEvent = {
      type: 'signal',
      targetingKey: signal.targetingKey ?? signal.rolloutId ?? 'fireweave-sdk',
      name: signal.name,
      properties: {
        kind: signal.kind,
        ...(signal.status !== undefined ? { status: signal.status } : {}),
        ...(signal.errorKind !== undefined ? { errorKind: signal.errorKind } : {}),
        ...(signal.message !== undefined ? { message: signal.message } : {}),
        ...(signal.value !== undefined ? { value: signal.value } : {}),
      },
    };
    if (signal.flagKey !== undefined) event.flagKey = signal.flagKey;
    if (signal.variant !== undefined) event.variant = signal.variant;
    if (signal.rolloutId !== undefined) event.rolloutId = signal.rolloutId;
    if (signal.changeId !== undefined) event.changeId = signal.changeId;
    if (signal.stampId !== undefined) event.stampId = signal.stampId;
    if (signal.timestamp !== undefined) event.timestamp = signal.timestamp;
    this.pending.push(event);
  }

  async flush(): Promise<void> {
    if (this.closed || !this.ready || this.pending.length === 0) return;
    const batch = this.pending.splice(0, this.pending.length);
    try {
      await this.requestJson(CAPTURE_PATH, { events: batch });
    } catch {
      // flush failures must not throw into shutdown paths; re-queue best-effort
      this.pending.unshift(...batch);
    }
  }

  async shutdown(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    try {
      const timeout = this.options.shutdownTimeoutMs ?? DEFAULT_ADAPTER_SHUTDOWN_TIMEOUT_MS;
      await Promise.race([
        this.flush(),
        new Promise<void>((resolve) => setTimeout(resolve, timeout)),
      ]);
    } catch {
      // never throw from shutdown
    }
    this.ready = false;
  }

  features(): AdapterRuntimeFeatures {
    return {
      remoteEvaluation: true,
      localEvaluation: false,
      localOnly: false,
      exposureEmission: true,
      sideEffectFreeReads: true,
      groupAnalytics: true,
    };
  }

  private toResolution(item: DecisionItem, quotaLimited: boolean): AdapterResolution {
    if (item.found === false) {
      const missing: AdapterResolution = { found: false };
      if (quotaLimited) missing.quotaLimited = true;
      return missing;
    }
    const resolution: AdapterResolution = {
      found: true,
      value: item.value,
      enabled: item.enabled ?? true,
    };
    if (item.variant != null) resolution.variant = item.variant;
    const reason = reasonToAdapter(item.reason);
    if (reason !== undefined) resolution.reason = reason;
    if (item.payload !== undefined) resolution.payload = item.payload;
    const meta = item.flagMetadata ?? {};
    if (typeof meta['fireweave.flagVersion'] === 'number') {
      resolution.version = meta['fireweave.flagVersion'];
    }
    if (typeof meta['fireweave.vendorFlagId'] === 'number') {
      resolution.vendorFlagId = meta['fireweave.vendorFlagId'];
    }
    if (typeof meta['fireweave.reasonCode'] === 'string') {
      resolution.reasonCode = meta['fireweave.reasonCode'];
    }
    if (quotaLimited || meta['fireweave.quotaLimited'] === true) {
      resolution.quotaLimited = true;
    }
    return resolution;
  }

  private async requestJson<T>(
    path: string,
    body: unknown,
    outerSignal?: AbortSignal,
  ): Promise<T> {
    const timeoutMs = this.options.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const onOuterAbort = () => controller.abort();
    if (outerSignal !== undefined) {
      if (outerSignal.aborted) controller.abort();
      else outerSignal.addEventListener('abort', onOuterAbort, { once: true });
    }
    try {
      let response: Awaited<ReturnType<FetchLike>>;
      try {
        response = await this.fetchImpl(`${this.apiUrl}${path}`, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            authorization: `Bearer ${this.apiKey}`,
          },
          body: JSON.stringify(body),
          signal: controller.signal,
        });
      } catch (err) {
        const aborted = err instanceof Error && err.name === 'AbortError';
        throw aborted ? new FireweaveError('Timeout') : new FireweaveError('Network');
      }
      if (response.status >= 400) {
        throw mapHttpStatus(response.status);
      }
      let parsed: unknown;
      try {
        parsed = await response.json();
      } catch {
        throw new FireweaveError('MalformedResponse');
      }
      return parsed as T;
    } finally {
      clearTimeout(timer);
      if (outerSignal !== undefined) {
        outerSignal.removeEventListener('abort', onOuterAbort);
      }
    }
  }
}
